### PR TITLE
feat(server): implement json.objkeys command (#104)

### PR DESCRIPTION
**Pull request by iko1**
Signed-off-by: iko1 <me@remotecpp.dev>

### CODE REVIEW

File: src/server/json_family.cc [hunk 1]
```diff
@@ -32,6 +32,7 @@ using OptBool = optional<bool>;
 using OptSizeT = optional<size_t>;
 using JsonReplaceCb = function<void(const string&, json&)>;
 using CI = CommandId;
+using StringVector = vector<string>;
```
Comment (romange) [inline src/server/json_family.cc:35]: common.h defines `StringVec` already.

File: src/server/json_family.cc [hunk 2]
```diff
@@ -490,8 +491,66 @@ OpResult<long> OpDel(const OpArgs& op_args, string_view key, string_view path) {
   return total_deletions;
 }
 
+OpResult<vector<StringVector>> OpObjKeys(const OpArgs& op_args, string_view key,
```
Comment (romange) [inline src/server/json_family.cc:494]: maybe worth adding a comment explaining what `vector<StringVector>` corresponds to.

Comment (romange) [inline src/server/json_family.cc:494]: maybe worth adding a comment explaining what `vector<StringVector>` corresponds to.

File: src/server/json_family.cc [hunk 3]
```diff
@@ -490,8 +491,66 @@ OpResult<long> OpDel(const OpArgs& op_args, string_view key, string_view path) {
...
+  }
+
+  vector<StringVector> vec;
+  auto cb = [&vec](const string_view& path, const json& val) {
+    if (!val.is_object()) {
+      vec.emplace_back();
+      return;
+    }
+
+    auto& current_object = vec.emplace_back();
```
Comment (romange) [inline src/server/json_family.cc:508]: what happens if you have a hierarchy of more than 2 levels?

File: src/server/json_family_test.cc [hunk 4]
```diff
@@ -514,4 +514,43 @@ TEST_F(JsonFamilyTest, Del) {
...
+  EXPECT_THAT(resp, ArgType(RespExpr::NIL_ARRAY));
+
+  resp = Run({"JSON.OBJKEYS", "json", "$.b"});
+  EXPECT_THAT(resp.GetVec(), ElementsAre("a"));
+
+  resp = Run({"JSON.OBJKEYS", "json", "$.*"});
+  ASSERT_THAT(resp, ArrLen(5));
+  const auto& arr = resp.GetVec();
+  EXPECT_THAT(arr[2].GetVec(), ElementsAre("a", "b"));
+  EXPECT_THAT(arr[3].GetVec(), ElementsAre("a", "b", "c"));
```
Comment (romange) [inline src/server/json_family_test.cc:535]: just to confirm OBJKEYS never goes beyond 2 levels in elasticache?

File: src/server/json_family_test.cc [hunk 5]
```diff
@@ -514,4 +514,43 @@ TEST_F(JsonFamilyTest, Del) {
...
+  resp = Run({"JSON.OBJKEYS", "json", "$.a"});
+  EXPECT_THAT(resp, ArgType(RespExpr::NIL_ARRAY));
+
+  resp = Run({"JSON.OBJKEYS", "json", "$.b"});
+  EXPECT_THAT(resp.GetVec(), ElementsAre("a"));
+
+  resp = Run({"JSON.OBJKEYS", "json", "$.*"});
+  ASSERT_THAT(resp, ArrLen(5));
+  const auto& arr = resp.GetVec();
+  EXPECT_THAT(arr[2].GetVec(), ElementsAre("a", "b"));
```
Comment (romange) [inline src/server/json_family_test.cc:534]: why skipped `arr[1]` and `arr[0]`?

File: src/server/json_family.cc (hunk 2 above)
Comment (romange) [reply src/server/json_family.cc:494]: after reading the code, I kinda understand the structure but it feels weird. 
the order of this vector is not well defined. Does `evaluate` preserves the original order?

**Review by romange — COMMENTED** [review]
Looks good. I just have a few questions about the command.

File: src/server/json_family_test.cc (hunk 5 above)
Comment (iko1) [reply src/server/json_family_test.cc:536]: I added the missing checks.

File: src/server/json_family_test.cc (hunk 4 above)
Comment (iko1) [reply src/server/json_family_test.cc:537]: It depends on what JsonPath you supply to the command.  I used JsonPath($.*) which traverses only one level of the hierarchy.
I added a UT that demonstrates recursive descent traversal ($..*) from level 2.

Comment (iko1) [reply src/server/json_family_test.cc:537]: I checked that my implementation is aligned with Elasticache and RedisJson. those implementations have some differences, one of them is: Elasticache prints an empty object and an item that is not an object as an empty array, and RedisJson differs between an empty object and an item that is not an object, an empty object is printed as an empty array and not object is printed as null.

File: src/server/json_family.cc (hunk 2 above)
Comment (iko1) [reply src/server/json_family.cc]: I did some checks and take a look at jsoncons codebase and what I can say is that evaluate preserves the hierarchy of the objects within the tree.

File: src/server/json_family.cc (hunk 3 above)
Comment (iko1) [reply src/server/json_family.cc:507]: The parent is always stored before their descendants in the Vector. A parent can have multiple descendants if the JsonPath is recursive and the parent index within the vector is smaller than the descendant index.

File: src/server/json_family_test.cc (hunk 4 above)
Comment (romange) [reply src/server/json_family_test.cc:537]: which flavor we are aligned with?

File: src/server/json_family.cc [hunk 6]
```diff
@@ -490,8 +490,68 @@ OpResult<long> OpDel(const OpArgs& op_args, string_view key, string_view path) {
...
+                                      JsonExpression expression) {
+  OpResult<json> result = GetJson(op_args, key);
+  if (!result) {
+    return result.status();
+  }
+
+  vector<StringVec> vec;
+  auto cb = [&vec](const string_view& path, const json& val) {
+    if (!val.is_object()) {
+      vec.emplace_back();
```
Comment (romange) [inline src/server/json_family.cc:505]: please add a comment here, saying it's aligned with elasticache flavor.

**Review by romange — COMMENTED** [review]
LGTM, please rebase and squash all changes into a single commit.

File: src/server/json_family_test.cc (hunk 4 above)
Comment (iko1) [reply src/server/json_family_test.cc:537]: We are aligned with Elasticache

**Review by romange — DISMISSED** [review]

**Review by romange — APPROVED** [review]

**Pull request merged**